File: zeroday-server/src/main/kotlin/com/zeroday/handler/handlers/GameEventHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.str
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

class GameEventHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.EVENT_LIST, RequestTypes.EVENT_JOIN, RequestTypes.EVENT_LEAVE,
        RequestTypes.EVENT_LEADERBOARD, RequestTypes.EVENT_SCORE, RequestTypes.EVENT_MY_EVENTS
    )

    override val requiresAuth: Boolean = true

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.EVENT_LIST -> listEvents(ctx)
            RequestTypes.EVENT_JOIN -> joinEvent(ctx, playerId, payload)
            RequestTypes.EVENT_LEAVE -> leaveEvent(ctx, playerId, payload)
            RequestTypes.EVENT_LEADERBOARD -> leaderboard(ctx, payload)
            RequestTypes.EVENT_SCORE -> updateScore(ctx, playerId, payload)
            RequestTypes.EVENT_MY_EVENTS -> myEvents(ctx, playerId)
            else -> HandlerResult.Err(ServerError.fromMessage("Unknown event request", ErrorCategory.VALIDATION))
        }
    }

    private fun JsonObject.strOr(key: String, default: String = ""): String =
        (this[key] as? JsonPrimitive)?.content ?: default

    private fun JsonObject.longOr(key: String, default: Long = 0L): Long =
        (this[key] as? JsonPrimitive)?.content?.toLongOrNull() ?: default

    private suspend fun listEvents(ctx: HandlerContext): HandlerResult {
        val events = ctx.services.gameEventService.getActiveEvents()
        return HandlerResult.Ok(ResponseTypes.EVENT_LIST, mapOf(
            "events" to events.map { ctx.services.gameEventService.serializeEvent(it) },
            "count" to events.size
        ))
    }

    private suspend fun joinEvent(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val eventId = payload.str("eventId")
            ?: return HandlerResult.Err(ServerError.fromMessage("eventId required", ErrorCategory.VALIDATION))
        val result = ctx.services.gameEventService.joinEvent(eventId, playerId)
        return result.fold(
            onSuccess = { event -> HandlerResult.Ok(ResponseTypes.EVENT_JOINED, ctx.services.gameEventService.serializeEvent(event)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Could not join event", ErrorCategory.VALIDATION)) }
        )
    }

    private suspend fun leaveEvent(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val eventId = payload.str("eventId")
            ?: return HandlerResult.Err(ServerError.fromMessage("eventId required", ErrorCategory.VALIDATION))
        val result = ctx.services.gameEventService.leaveEvent(eventId, playerId)
        return result.fold(
            onSuccess = { HandlerResult.Ok(ResponseTypes.EVENT_LEFT, mapOf("eventId" to eventId)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Could not leave event", ErrorCategory.VALIDATION)) }
        )
    }

    private suspend fun leaderboard(ctx: HandlerContext, payload: JsonObject): HandlerResult {
        val eventId = payload.str("eventId")
            ?: return HandlerResult.Err(ServerError.fromMessage("eventId required", ErrorCategory.VALIDATION))
        val participants = ctx.services.gameEventService.getEventLeaderboard(eventId)
        return HandlerResult.Ok(ResponseTypes.EVENT_LEADERBOARD, mapOf(
            "eventId" to eventId,
            "participants" to participants.mapIndexed { idx, p -> mapOf(
                "playerName" to p.playerName,
                "score" to p.score,
                "rank" to (idx + 1)
            )}
        ))
    }

    private suspend fun updateScore(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val eventId = payload.str("eventId")
            ?: return HandlerResult.Err(ServerError.fromMessage("eventId required", ErrorCategory.VALIDATION))
        val delta = payload.longOr("delta")
        if (delta <= 0) return HandlerResult.Err(ServerError.fromMessage("delta must be positive", ErrorCategory.VALIDATION))

        val result = ctx.services.gameEventService.updateScore(eventId, playerId, delta)
        return result.fold(
            onSuccess = { HandlerResult.Ok(ResponseTypes.EVENT_SCORE_UPDATED, mapOf("eventId" to eventId, "playerId" to playerId, "delta" to delta)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Score update failed", ErrorCategory.VALIDATION)) }
        )
    }

    private suspend fun myEvents(ctx: HandlerContext, playerId: String): HandlerResult {
        val events = ctx.services.gameEventService.getPlayerEvents(playerId)
        return HandlerResult.Ok(ResponseTypes.EVENT_MY_EVENTS, mapOf(
            "events" to events.map { ctx.services.gameEventService.serializeEvent(it) },
            "count" to events.size
        ))
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/GameEventHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class GameEventHandler : IHandler
{
    private readonly IPlayerService _playerService;

    public string MessageType => "event";

    public GameEventHandler(IPlayerService playerService)
    {
        _playerService = playerService;
    }

    public Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "Missing payload" }));

        var action = payload.Value.Str("action");
        return action switch
        {
            "list" => HandleList(connectionId),
            "join" => HandleJoin(connectionId, payload.Value),
            "leave" => HandleLeave(connectionId, payload.Value),
            "leaderboard" => HandleLeaderboard(connectionId, payload.Value),
            "score" => HandleScore(connectionId, payload.Value),
            "my_events" => HandleMyEvents(connectionId, payload.Value),
            _ => Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = $"Unknown event action: {action}" }))
        };
    }

    private static Task<IActionResult> HandleList(string connectionId)
    {
        return Task.FromResult<IActionResult>(
            new MessageResult("event_list", connectionId, new
            {
                events = Array.Empty<object>(),
                count = 0
            }));
    }

    private Task<IActionResult> HandleJoin(string connectionId, JsonElement payload)
    {
        var eventId = payload.Str("eventId");
        if (string.IsNullOrEmpty(eventId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "eventId required" }));

        return Task.FromResult<IActionResult>(
            new MessageResult("event_joined", connectionId, new { eventId }));
    }

    private Task<IActionResult> HandleLeave(string connectionId, JsonElement payload)
    {
        var eventId = payload.Str("eventId");
        if (string.IsNullOrEmpty(eventId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "eventId required" }));

        return Task.FromResult<IActionResult>(
            new MessageResult("event_left", connectionId, new { eventId }));
    }

    private static Task<IActionResult> HandleLeaderboard(string connectionId, JsonElement payload)
    {
        var eventId = payload.Str("eventId");
        if (string.IsNullOrEmpty(eventId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "eventId required" }));

        return Task.FromResult<IActionResult>(
            new MessageResult("event_leaderboard", connectionId, new
            {
                eventId,
                participants = Array.Empty<object>()
            }));
    }

    private static Task<IActionResult> HandleScore(string connectionId, JsonElement payload)
    {
        var eventId = payload.Str("eventId");
        var delta = payload.Long("delta");

        if (string.IsNullOrEmpty(eventId))
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "eventId required" }));
        if (delta is null || delta <= 0)
            return Task.FromResult<IActionResult>(new MessageResult("error", connectionId, new { message = "delta must be positive" }));

        var playerId = payload.Str("playerId");

        return Task.FromResult<IActionResult>(
            new MessageResult("event_score_updated", connectionId, new
            {
                eventId,
                playerId,
                delta
            }));
    }

    private static Task<IActionResult> HandleMyEvents(string connectionId, JsonElement payload)
    {
        return Task.FromResult<IActionResult>(
            new MessageResult("event_my_events", connectionId, new
            {
                events = Array.Empty<object>(),
                count = 0
            }));
    }
}
